### PR TITLE
Test1.2

### PR DESCRIPTION
**Pull request by sudarson-25**
PR created for review

### CODE REVIEW

File: Training/Training/testcomplex.c [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+// Program to perform arithmetic operations on complex numbers.
+// Sudarson S
+// ------------------------------------------------------------------------------------------------
+
+#include<stdio.h>
+#include"complex.h"
+
+/// <summary>Runs testcases for various arithmetic operations</summary>
+static void TestCases () {
+   Complex input[] = { 5,4,6,-5 }, expected[] = { 11,-1,-1,9,50,-1,5,-4 };
```
Comment (sanjivtmm) [inline Training/Training/testcomplex.c]: You should do the initialization like this:
   Complex input[] = { {5,4},{6,-5} }, expected[] = { {11,-1},{-1,9},{50,-1},{5,-4} };
It clearly shows how many complex numbers are being initialized and what are the real and imaginary parts.

File: Training/Training/testcomplex.c [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+   double modExpected = 6.4031242374328485;
+   Complex D = OpAdd (input[0], input[1]), E = OpSub (input[0], input[1]),
```
Comment (sanjivtmm) [inline Training/Training/testcomplex.c]: Readability is not good. Improve it by formatting your code as follows:
   Complex  D = OpAdd (input[0],input[1]),
                   E = OpSub (input[0], input[1]),
                   F = OpMul (input[0], input[1]), G = OpConj (input[0]);

File: Training/Training/testcomplex.c [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      F = OpMul (input[0], input[1]), G = OpConj (input[0]);
+   D.real == expected[0].real && D.img == expected[0].img ? printf ("\nADD Passed") : printf ("\nADD Failed");
```
Comment (sanjivtmm) [inline Training/Training/testcomplex.c]: Use parenthesis to improve readability.
The tertiary operator is better used when it returns a value and can be used in an expression.
A regular if..else statement is better in this case.

File: Training/Training/testcomplex.c [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+/// <summary>Prints a complex number</summary>
+static void PrintComplex (Complex C) {
+   C.img >= 0 ? printf ("%0.2f+%0.2fi\n", C.real, C.img) : printf ("%0.2f%0.2fi\n", C.real, C.img);
+}
+
+int main () {
+   Complex A = { 0,0 }, B = { 0,0 };
+   printf ("Enter the real part of the first number: ");
```
Comment (sanjivtmm) [inline Training/Training/testcomplex.c]: Why are you still manually testing the code? This is not very robust. Did you try entering a non-numeric string, or a string of numbers like 1.2 2.4 3.6 4.8?

File: Training/Training/testcomplex.c [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+   printf ("\nAddition: ");
+   PrintComplex (OpAdd (A, B));
+   printf ("\nSubtraction: ");
+   PrintComplex (OpSub (A, B));
+   printf ("\nMultiplication: ");
+   PrintComplex (OpMul (A, B));
+   printf ("\nModulus of A: %0.2lf\n", OpMod (A));
+   printf ("\nConjugate of A: ");
+   PrintComplex (OpConj (A));
+   TestCases ();
```
Comment (sanjivtmm) [inline Training/Training/testcomplex.c:58]: Too few automated tests.
It is possible to the function TestCases to take parameters.
Think of how to automatically generate the test cases.
This is a thought for the future.

**Comment by Dakshita-Minocha** [comment]
Branch is named correctly but PR should always be named with the original name i.e., "Test1" and be provided with a description. 
"Test1: Complex Numbers" in this case.

File: Training/Training/complex.c [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+   return A;
+}
+
+void PrintComplex (Complex C) {
+   C.Img >= 0 ? printf ("%0.2f +%0.2fi\n", C.Real, C.Img) : printf ("%0.2f %0.2fi\n", C.Real, C.Img);
+}
+
+bool Compare (Complex C, Complex D) {
+   return C.Real == D.Real && C.Img == D.Img;
+}
```
Comment (Dakshita-Minocha) [inline Training/Training/complex.c]: Conventionally, we name methods like this (that return only bool) IsEqual (). Compare (a, b) methods return 1, 0, -1 as value based on whether a > b, a = b or a < b.

File: Training/Training/testcomplex.c [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+         J = { input[i].Real * input[i + 1].Real - input[i].Img * input[i + 1].Img,
+         input[i].Real * input[i + 1].Img + input[i].Img * input[i + 1].Real },
+         K = { input[i].Real, -input[i].Img };
+      printf (Compare (D, H) ? "\nADD Passed" : "\nADD Failed");
+      printf (Compare (E, I) ? "\nSUB Passed" : "\nSUB Failed");
+      printf (Compare (F, J) ? "\nMUL Passed" : "\nMUL Failed");
+      ////Testing CONJ and MOD with the first complex number
+      printf (Compare (G, K) ? "\nCONJ Passed" : "\nCONJ Failed");
+      printf (OpMod (input[i]) == sqrt (input[i].Real * input[i].Real + input[i].Img * input[i].Img) ?
+         "\nMOD Passed\n" : "\nMOD Failed\n");
```
Comment (Dakshita-Minocha) [inline Training/Training/testcomplex.c]: I think the best way for you to indicate a failed test would be to print failed (or passed) only once for continue;
```suggestion
      printf ("Case 1: ");
      if (!Compare (D, H)) { printf("\nADD Failed"); continue; }
     // etc
     printf ("Passed");
```
and so on. This has two advantages: 1. It fails and stop running that particular test the moment it fails (which is desirable) and it tells you exactly where it is failing. There is a thing called "[Fail fast and Fail Loud](https://stackoverflow.com/questions/2807241/what-does-the-expression-fail-early-mean-and-when-would-you-want-to-do-so)"
2. It declutters and helps the user understand where the problem is occurring in a minimum number of words.

**Review by Dakshita-Minocha — CHANGES_REQUESTED** [review]

File: Training/Training/testcomplex.c (hunk 7 above)
Comment (Dakshita-Minocha) [reply Training/Training/testcomplex.c]:
```suggestion
      printf (Compare (Op_Add (a, b) , H) ? "\nADD Passed" : "\nADD Failed");
```
This is more intuitive. The cache-ing is not required as the value is only being used once.

**Review by Dakshita-Minocha — CHANGES_REQUESTED** [review]

**Review by Dakshita-Minocha — APPROVED** [review]

**Pull request closed without merging**